src/Agent.Worker/Handlers/Handler.cs

```diff
@@ -298,24 +300,65 @@ protected void AddPrependPathToEnvironment()
             }
         }
 
-        protected void RemovePSModulePathFromEnvironment()
+        [SupportedOSPlatform("windows")]
+        protected bool PsModulePathContainsPowershellCoreLocations()
         {
-            if (AgentKnobs.CleanupPSModules.GetValue(ExecutionContext).AsBoolean() && PlatformUtil.RunningOnWindows)
+            bool checkLocationsKnob = AgentKnobs.CheckPsModulesLocations.GetValue(HostContext).AsBoolean();
+
+            bool isPwshCore = Inputs.TryGetValue("pwsh", out string pwsh) && StringUtil.ConvertToBoolean(pwsh);
+
+            if (!PlatformUtil.RunningOnWindows || !checkLocationsKnob || isPwshCore)
             {
-                bool useInteropKnob = AgentKnobs.UseInteropToFindParentProcess.GetValue(ExecutionContext).AsBoolean();
+                return false;
+            }
+
+            const string PSModulePath = nameof(PSModulePath);
+
+            bool localVariableExists = Environment.TryGetValue(PSModulePath, out string localVariable);
+            bool localVariableContainsPwshLocations = PsModulePathUtil.ContainsPowershellCoreLocations(localVariable);
+
+            // Special case when the env variable is set for local process environment
+            // for example by vso command in a preceding pipeline step
+            if (localVariableExists && !localVariableContainsPwshLocations)
+            {
+                return false;
+            }
+
+            string systemVariable = System.Environment.GetEnvironmentVariable(PSModulePath);
```

Comment:
I don't see a difference between "systemVariable" and "localVariable".  They seem to check the same env variable (PSModulePath) the same way.

Reply:
When a variable such as PsModulePath is added via `setvariable` vso command, for example, in a preceding pipeline step, it is added to a dictionary of local environment variables without affecting the system variable. When a handler starts a new process, the variables from this dictionary are passed to it. 

The idea is that if there's already a value in the local environment dictionary, most likely it is deliberately put there by user and we don't have to check the system env variable.

Reply:
This might be too convoluted and if it is not worth it to try to account for it, please let me know and I will remove it